flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1103,23 +1096,64 @@ private void archiveExecutionGraph(ExecutionGraphInfo executionGraphInfo) {
                     executionGraphInfo.getArchivedExecutionGraph().getJobID(),
                     e);
         }
+    }
 
-        // do not create an archive for suspended jobs, as this would eventually lead to multiple
-        // archive attempts which we currently do not support
-        if (executionGraphInfo.getArchivedExecutionGraph().getState().isGloballyTerminalState()) {
-            final CompletableFuture<Acknowledge> executionGraphFuture =
-                    historyServerArchivist.archiveExecutionGraph(executionGraphInfo);
-
-            executionGraphFuture.whenComplete(
-                    (Acknowledge ignored, Throwable throwable) -> {
-                        if (throwable != null) {
-                            log.info(
-                                    "Could not archive completed job {}({}) to the history server.",
-                                    executionGraphInfo.getArchivedExecutionGraph().getJobName(),
-                                    executionGraphInfo.getArchivedExecutionGraph().getJobID(),
-                                    throwable);
-                        }
-                    });
+    private CompletableFuture<Acknowledge> archiveExecutionGraphToHistoryServer(
+            ExecutionGraphInfo executionGraphInfo) {
+
+        final CompletableFuture<Acknowledge> executionGraphFuture =
+                FutureUtils.orTimeout(
+                        historyServerArchivist.archiveExecutionGraph(executionGraphInfo),
+                        configuration.get(ClusterOptions.CLUSTER_SERVICES_SHUTDOWN_TIMEOUT),
+                        TimeUnit.MILLISECONDS,
+                        getMainThreadExecutor());
```

Comment:
In our production we met a scenario that the user specifies `historyserver.archive.fs.dir` to a remote OSS path. It takes a long time to archive the files to OSS. If a timeout happens, it means that the storage is not reliable. I'm worried that if no timeout is added, the job will never end in the worst scenario. I fully understand that if a user enables the configs related to the history server, he/she wants to have the archiving finished. However, if the worst case happened, he/she has to shutdown the job manually, with no resource clean up. In my opinion, resource clean up is more important than archiving, even when the user enables the history server. Thus, I think a timeout is necessary for the archiving.

Reply:
But wouldn't that be an archiving issue? Looking into `FsJobArchivist.archiveJob` I don't see why we need to list the files of the bucket? Where is this expensive operation added? `archiveJob` defines a path that it writes the JSON file to. No iteration over file objects or anything. That shouldn't cause any problems in object stores, should it? Even the [delete in FsJobArchivist:87](https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java#L87) is not recursive and shouldn't be too expensive since it's operating on a distinct object key (i.e. the path).

To me, it sounds like an issue of the FileSystem implementation. Or am I missing something here?

Reply:
> To me, it sounds like an issue of the FileSystem implementation. Or am I missing something here?

Yes, you are right. It's an issue of the `FileSystem` implementation. The operation of listing the files in the bucket is the detail of the OSS `FileSystem` implementation. However, we could not make sure all the implementations of the `FileSystem` has the timeout mechanism. If not, I think maybe it could be safer if we add a timeout here. We try best to archive the ExecutionGraph, but once it takes a long time, we could output an error to tell the users the destination they specifies is unreachable within limited time.

Reply:
I thought it once more. Then I realize that I shouldn't couple the issue of the `FileSystem` implementation with this issue. I will remove the timeout here. Also the test case `testArchiveFailedWhenTimeoutExceeded` will be removed ,too.